flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoder.java

```diff
@@ -0,0 +1,143 @@
+
+import org.apache.flink.runtime.io.network.NetworkClientHandler;
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelHandlerContext;
+import org.apache.flink.shaded.netty4.io.netty.channel.ChannelInboundHandlerAdapter;
+
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.FRAME_HEADER_LENGTH;
+import static org.apache.flink.runtime.io.network.netty.NettyMessage.MAGIC_NUMBER;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * Decodes messages from the fragmentary netty buffers. This decoder assumes the
+ * messages have the following format:
+ * +-----------------------------------+--------------------------------+
+ * | FRAME_HEADER ||  MESSAGE_HEADER   |     DATA BUFFER (Optional)     |
+ * +-----------------------------------+--------------------------------+
+ *
+ * This decoder decodes the frame header and delegates the other following work
+ * to corresponding message parsers according to the message type. During the process
+ * of decoding, the decoder and parsers try best to eliminate copying. For the frame
+ * header and message header, it only cumulates data when they span multiple input buffers.
+ * For the buffer part, it copies directly to the input channels to avoid future copying.
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class NettyMessageClientDecoder extends ChannelInboundHandlerAdapter {
+
+	/** The message parser for buffer response. */
+    private final NettyMessageDecoderDelegate bufferResponseDecoderDelegate;
+
+    /** The message parser for other messages other than buffer response. */
+	private final NettyMessageDecoderDelegate nonBufferResponseDecoderDelegate;
+
+	/** The cumulation buffer for the frame header part. */
+	private ByteBuf frameHeaderBuffer;
+
+	/**
+	 * The chosen message parser for the current message. If it is null, then
+	 * we are decoding the frame header part, otherwise we are decoding the actual
+	 * message.
+	 */
+	private NettyMessageDecoderDelegate currentDecoderDelegate;
+
+    NettyMessageClientDecoder(NetworkClientHandler networkClientHandler) {
+        this.bufferResponseDecoderDelegate = new BufferResponseDecoderDelegate(networkClientHandler);
+        this.nonBufferResponseDecoderDelegate = new NonBufferResponseDecoderDelegate();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        bufferResponseDecoderDelegate.onChannelActive(ctx.alloc());
+        nonBufferResponseDecoderDelegate.onChannelActive(ctx.alloc());
+
+		frameHeaderBuffer = ctx.alloc().directBuffer(FRAME_HEADER_LENGTH);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
```

Comment:
Do we also need to release the resources via `exceptionCaught` interface method as did in other handler?

Reply:
I think it might be ok to not deal with `exceptionCaught` explicitly, since `channelInactive` is called when channel get closed, and in `CreditBasedPartitionClientHander`, it closes the channel in `exceptionCaught`, thus it will close the channel. Besides, I think we should always be able to close the channel, otherwise we should have channel leakage problem.

Reply:
It might be better to handle them dependently, not relying on the implicit logic of other handlers.